File: Python/72.编辑距离.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        if len(word1) == 0 or len(word2) == 0:
            return len(word1) + len(word2)
        dp: list[int] = [[0] * (1 + len(word2)) for _ in range(1 + len(word1))]
        for i in range(len(word1) + 1):
            dp[i][0] = i
        for j in range(len(word2) + 1):
            dp[0][j] = j
        for i in range(1, len(word1) + 1):
            for j in range(1, len(word2) + 1):
                if word1[i - 1] == word2[j - 1]:
                    dp[i][j] = dp[i - 1][j - 1]
                else:
                    dp[i][j] = 1 + min(dp[i - 1][j], min(dp[i][j - 1], dp[i - 1][j - 1]))
        # return dp[-1][-1]
        return dp[len(word1)][len(word2)]

File: TypeScript/72.编辑距离.ts
function minDistance(word1: string, word2: string): number {
    if (word1.length == 0 || word2.length == 0)
        return word1.length + word2.length
    const dp = [...new Array(1 + word1.length)].map(_ => new Array(1 + word2.length).fill(0))
    for (let i = 0; i <= word1.length; ++i)
        dp[i][0] = i
    for (let j = 0; j <= word2.length; ++j)
        dp[0][j] = j
    for (let i = 1; i <= word1.length; ++i)
        for (let j = 1; j <= word2.length; ++j)
            if (word1[i-1] == word2[j-1])
                dp[i][j] = dp[i-1][j-1]
            else
                dp[i][j] = 1 + Math.min(dp[i-1][j], Math.min(dp[i][j-1], dp[i-1][j-1]))
    return dp.pop().pop()
    // return dp[word1.length][word2.length]
};
